src/dev/lupio/lupio_blk.cc

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_blk.hh"
+
+#include "debug/LupioBLK.hh"
+#include "dev/dma_device.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioBLK.hh"
+
+/* Shared fields for CTRL and STAT registers */
+#define LUPIO_BLK_TYPE  0x2
+
+/* Specific fields for CTRL */
+#define LUPIO_BLK_IRQE  0x1
+
+/* Specific fields for STAT */
+#define LUPIO_BLK_BUSY  0x1
+#define LUPIO_BLK_ERRR  0x4
+
+#define SECTOR_BITS   9
+#define SECTOR_SIZE   (1ULL << SECTOR_BITS)
+
+namespace gem5
+{
+LupioBLK::LupioBLK(const Params &params) :
+    DmaDevice(params),
+    platform(*params.platform),
+    dmaEvent([this]{ dmaEventDone(); }, name()),
+    pioAddr(params.pio_addr),
+    pioSize(params.pio_size),
+    image(*params.image),
+    lupioBLKIntID(params.int_id)
+{
+    static_assert(SECTOR_SIZE == SectorSize, "Sector size of disk image must"
+                    " match LupIO device");
+    gem5_assert(image.size() <= pow(2, 32));    //Max number of blocks
+    nbBlocks = image.size();
+    DPRINTF(LupioBLK, "LupioBLK initalized\n");
+}
+
+void
+LupioBLK::dmaEventDone()
+{
+    int64_t _offset = lba;
+
+    if (writeOp) {
+        for (int i = 0; i < reqLen; i += SECTOR_SIZE){
+            image.write(&reqData[i], _offset);
+            _offset++;
```

Comment:
- You could potentially increment the offset as part of the for loop itself: `for (int i = 0; i < reqLen; i += SECTOR_SIZE, offset++) {`
- Or, maybe even better, you can drop `_offset` and just use `i` added to `lba`: `image.write(&reqData[i], lba + i);`